### PR TITLE
Bed files intersection fixes

### PR DESCRIPTION
**Pull request by cansavvy**
#### What scientific question is your analysis addressing?
The BED files we use for TMB calculation denominators seemed too big (#565). After looking into the bedtools documentation, we realized it wasn't doing *exactly* what we needed. I've added sort and merge steps and also have made the intersections for greater than 2 files happen in their own individual steps. (See #564)

#### What was your approach?
I've added sorts and merges to get rid of the redundancy noted in #565 

#### What GitHub issue does your pull request address?
#564 

#### Which areas should receive a particularly close look?
- Are there other quality checks I should do to test if these BED files are more sensible? They are smaller and have no overlaps. 

#### Is there anything that you want to discuss further?
- Do these BED files make more sense? Are all the issues we know of addressed? 

#### Is the analysis in a mature enough form that the resulting figure(s) and/or table(s) are ready for review?
No. Results yet. 

#### What is your summary of the results?
I need to re-run this in AWS after it is decided it makes sense. I'll have another PR with all the updated plots and notebooks with the data I pull from AWS. 

#### Reproducibility Checklist
- [x] The dependencies required to run the code in this pull request have been added to the project Dockerfile.
- [x] This analysis has been added to continuous integration.

#### Documentation Checklist
Although the BED files have changed, their names are the same 
- [x] This analysis module has a `README` and it is up to date.
- [x] This analysis is recorded in the table in `analyses/README.md` and the entry is up to date.
- [x] The analytical code is documented and contains comments.

### CODE REVIEW

File: analyses/snv-callers/run_caller_consensus_analysis-tcga.sh [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+###################### Create intersection BED files ###########################
+# Convert GTF to BED file for use in bedtools
+# Here we are only extracting lines with as a CDS i.e. are coded in protein
+gunzip -c data/gencode.v27.primary_assembly.annotation.gtf.gz \
+  | awk '$3 ~ /CDS/' \
+  | convert2bed --do-not-sort --input=gtf - \
+  > scratch/coding_sequences.bed
+
+# Sort so we can merge
+bedtools sort -i coding_sequences.bed |
```
Comment (jashapiro) [inline analyses/snv-callers/run_caller_consensus_analysis-tcga.sh]: Using unix sort is faster (probably not a concern, but it is what bedtools manual recommends, see the "Disclaimer" https://bedtools.readthedocs.io/en/latest/content/tools/sort.html#optional-sorting-behavior )
```suggestion
sort -k 1,1 -k 2,2n coding_sequences.bed | 
```

File: analyses/snv-callers/run_caller_consensus_analysis-tcga.sh [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+# Make WXS coding BED file
+# TODO: This file path will need to be updated when the TCGA target BED file is 
+# added to the data release
+bedtools intersect \
+  -a analyses/snv-callers/ref_files/gencode.v19.basic.exome.hg38liftover.bed  \
+  -b $cds_file \
+  > scratch/coding_wxs.bed
+
+# Sort so we can merge
+bedtools sort -i scratch/coding_wxs.bed |
```
Comment (jashapiro) [inline analyses/snv-callers/run_caller_consensus_analysis-tcga.sh]: Same suggestion as above.
```suggestion
sort -k 1,1 -k 2,2n scratch/coding_wxs.bed | 
```

File: analyses/snv-callers/run_caller_consensus_analysis-tcga.sh [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+
+# Sort so we can merge
+bedtools sort -i coding_sequences.bed | 
+bedtools merge -i stdin  > $cds_file
+
+# Make WXS coding BED file
+# TODO: This file path will need to be updated when the TCGA target BED file is 
+# added to the data release
+bedtools intersect \
+  -a analyses/snv-callers/ref_files/gencode.v19.basic.exome.hg38liftover.bed  \
```
Comment (jashapiro) [inline analyses/snv-callers/run_caller_consensus_analysis-tcga.sh]: Did you check whether it made sense to sort and merge this file first as well? Or is it already free of overlaps?

Comment (cansavvy) [reply analyses/snv-callers/run_caller_consensus_analysis-tcga.sh]: I will double check. It turns out its not the right BED file anyway (No issue to cite but something we've heard from d3b folks).

Comment (cansavvy) [reply analyses/snv-callers/run_caller_consensus_analysis-tcga.sh]: There's no overlap, so we can continue to use this file this way for now.

**Comment by cansavvy** [comment]
Alrighty @jashapiro since you've reviewed most of this already anyway, any other comments?

File: analyses/snv-callers/run_caller_consensus_analysis-pbta.sh [hunk 4]
```diff
@@ -49,31 +49,55 @@ python3 analyses/snv-callers/scripts/01-setup_db.py \
...
 # Convert GTF to BED file for use in bedtools
 # Here we are only extracting lines with as a CDS i.e. are coded in protein
 gunzip -c data/gencode.v27.primary_assembly.annotation.gtf.gz \
   | awk '$3 ~ /CDS/' \
   | convert2bed --do-not-sort --input=gtf - \
-  > $cds_file
+  > scratch/coding_sequences.bed
+
+# Sort so we can merge
+sort -k 1,1 -k 2,2n coding_sequences.bed |
```
Comment (jashapiro) [inline analyses/snv-callers/run_caller_consensus_analysis-pbta.sh]: The output and inputs here don't match (scratch/coding_sequences.bed vs. coding_sequences.bed). Might be worth just piping the whole way

```gunzip -c data/gencode.v27.primary_assembly.annotation.gtf.gz \
  | awk '$3 ~ /CDS/' \
  | convert2bed --do-not-sort --input=gtf - \
  | sort -k 1,1 -k 2,2n \
  | bedtools merge -i stdin  \
  > $cds_file
```

File: analyses/snv-callers/run_caller_consensus_analysis-pbta.sh [hunk 5]
```diff
@@ -49,31 +49,55 @@ python3 analyses/snv-callers/scripts/01-setup_db.py \
...
+
+# Intersect the mutect and strelka coding beds into one
+bedtools intersect \
+  -a scratch/wgs_coding_mutect.bed \
+  -b scratch/wgs_coding_strelka.bed \
+   > scratch/wgs_coding_strelka_mutect.bed
+
+# Merge these ranges into one
+sort -k 1,1 -k 2,2n scratch/wgs_coding_strelka_mutect.bed | 
+bedtools merge -i stdin > $coding_wgs_bed
```
Comment (jashapiro) [inline analyses/snv-callers/run_caller_consensus_analysis-pbta.sh]: Same suggestion as before: pipe all the things
```suggestion
bedtools intersect \
  -a scratch/wgs_coding_mutect.bed \
  -b scratch/wgs_coding_strelka.bed \
  | sort -k 1,1 -k 2,2n \
  | bedtools merge \
  > $coding_wgs_bed
```

File: analyses/snv-callers/run_caller_consensus_analysis-pbta.sh [hunk 6]
```diff
@@ -49,31 +49,55 @@ python3 analyses/snv-callers/scripts/01-setup_db.py \
...
+   
+##################### Make WXS coding BED file
+# Intersect coding and WXS ranges, sort and merge 
+bedtools intersect \
+  -a data/WXS.hg38.100bp_padded.bed  \
+  -b $cds_file |
+sort -k 1,1 -k 2,2n -i stdin |
+bedtools merge -i stdin > $coding_wxs_bed
```
Comment (jashapiro) [inline analyses/snv-callers/run_caller_consensus_analysis-pbta.sh]:
```suggestion
bedtools intersect \
  -a data/WXS.hg38.100bp_padded.bed  \
  -b $cds_file \
  | sort -k 1,1 -k 2,2n  \
  | bedtools merge \
  > $coding_wxs_bed
```

File: analyses/snv-callers/run_caller_consensus_analysis-pbta.sh (hunk 4 above)
Comment (cansavvy) [reply analyses/snv-callers/run_caller_consensus_analysis-pbta.sh]: This did work, btw.

File: analyses/snv-callers/run_caller_consensus_analysis-tcga.sh [hunk 7]
```diff
@@ -57,15 +57,22 @@ python3 analyses/snv-callers/scripts/01-setup_db.py \
...
+# TODO: Update this BED when we get the target BEDs updated in v15/v16
 bedtools intersect \
   -a analyses/snv-callers/ref_files/gencode.v19.basic.exome.hg38liftover.bed  \
   -b $cds_file \
-  > $coding_wxs_bed
+  > scratch/coding_wxs.bed
+
+# Sort so we can merge
+sort -k 1,1 -k 2,2n scratch/coding_wxs.bed | 
+bedtools merge -i stdin  > $coding_wxs_bed
```
Comment (jashapiro) [inline analyses/snv-callers/run_caller_consensus_analysis-tcga.sh]:
```
bedtools intersect \
  -a analyses/snv-callers/ref_files/gencode.v19.basic.exome.hg38liftover.bed  \
  -b $cds_file \
  | sort -k 1,1 -k 2,2n \
  | bedtools merge \
  > $coding_wxs_bed
```

**Pull request merged**